### ISSUE TITLE
VIP: Add License Header

### ISSUE THREAD

**Issue by fubuloubu** [vyperlang/vyper#1955]
## Simple Summary
Allow specifying an SPDX-compatable license in the header of a file (similar to `version`)

## Motivation
Sometimes individual files can have a different license than the wider system, so we should allow a shorthand way to specify that

## Specification
If defined, this must be near the top of a file:
```python
# @license SPDX_ID [SPDX_MODIFIER SPDX_ID]*
```

## Backwards Compatibility
No backwards incompatibilities.

## Dependencies
No dependencies.

## References
https://en.wikipedia.org/wiki/Software_Package_Data_Exchange

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

**Comment by fubuloubu** [comment]
Meeting notes: Move to Contract-level natspec docstring

**Comment by arjunaskykok** [comment]
"If defined, this **must be near the top of a file**:" -> Do we really need to enforce this? Right now in Solidity (0.6.8), you can put SPDX on the bottom of the file. The compiler doesn't care as long as you define it.

**Comment by iamdefinitelyahuman** [comment]
Because we want it in the natspec, it has to go at the very top of the contract to be correctly parsed as a docstring.

**Comment by fubuloubu** [comment]
@iamdefinitelyahuman did we ever track a decision to move this (and versioning) to NatSpec?

**Comment by fubuloubu** [comment]
@arjunaskykok maybe join us in #1996 and we can discuss!

**Comment by arjunaskykok** [comment]
@fubuloubu Ah, sorry. I'll join the meetup next one!

**Comment by arjunaskykok** [comment]
@iamdefinitelyahuman So something like this? Then we should update the specification on this issue.
```python
# @version 0.2.0b13
"""
@license SPDX_ID GPL-v3.0
@title A greeting program
@author Arjuna Sky Kok
@notice You can use this contract for greeting
@dev Greetings must be done in a polite way
"""
greeting: bytes[20]

@public
def __init__():
    self.greeting = "Hello World"
```

**Comment by arjunaskykok** [comment]
Also, when compiling source code, in order to use NatSpec, Vyper uses devdoc and userdoc flags. Nah, if we want to put a license in NatSpec, does it mean we **don't check** the existence of the license when we **don't** use devdoc and userdoc flags?

In other words, is a license optional or a must in Vyper source code? Right now, in Solidity, it's a must. You must add a license in Solidity source code.

Reference: https://solidity.readthedocs.io/en/v0.6.9/layout-of-source-files.html#spdx-license-identifier

**Comment by fubuloubu** [comment]
@arjunaskykok yes, except remove the word `SPDX_ID` from your example (`GPL-3.0-only` would be the correct [`SPDX_ID`](https://spdx.org/licenses/) instead of `GPL-v3.0`)

Note: `SPDX_MODIFIER` refers to one of `AND` or `OR`, as a way to combine licenses.

**Comment by fubuloubu** [comment]
@arjunaskykok I think the intention was that this might be standardized in languages as a means of locating files who have licenses that do not match their corresponding projects, or supercede them. I think it is the intent of frameworks that make use of these license identifiers to actually validate and structure them.

This was primarily being thought of in regards to ethPM v3, which has an optional license field on every source object (in addition to the project-wide one)

**Comment by fubuloubu** [comment]
Fixed in #2007
